### PR TITLE
Deletes unecessary prediction files

### PR DESCRIPTION
**Pull request by gui-miotto**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -12,7 +12,7 @@ joblib
 psutil
 pyyaml
 liac-arff
-pandas<1.0
+pandas
```
Comment (gui-miotto) [inline requirements.txt]: I don't know why this restriction was still here. Maybe an issue while merging?

File: autosklearn/ensemble_builder.py [hunk 2]
```diff
@@ -671,12 +679,32 @@ def predict(self, set_: str,
...
-            model_file_id = model_file.split('.')[-2]
-            if model_file_id not in cand_ids:
-                os.remove(model_file)
-                self.logger.info("Deleted file of non-candidate model %s", model_file)
+        for pred_file in self.read_preds.keys():
+            if pred_file not in candidates and pred_file not in self.deleted_preds:
+                # Delete prediction file
+                try:
+                    pred_lock = pred_file + '.lock'
+                    with lockfile.LockFile(pred_lock):
```
Comment (gui-miotto) [inline autosklearn/ensemble_builder.py]: I added the locks, but I'm not quite sure that they are necessary. 
Before adding the locks, I ran the test_delete_non_candidate_models with very loose time constraints and got no errors.

File: requirements.txt (hunk 1 above)
Comment (mfeurer) [reply requirements.txt]: Me neither. As long as it works it's fine I guess.

File: autosklearn/ensemble_builder.py (hunk 2 above)
Comment (mfeurer) [reply autosklearn/ensemble_builder.py]: I'm not sure either, but let's better be safe than sorry.

File: autosklearn/ensemble_builder.py [hunk 3]
```diff
@@ -671,12 +679,32 @@ def predict(self, set_: str,
...
     def _delete_non_candidate_models(self, candidates):
-        cand_ids = [cand.replace('_', '.').split('.')[-2] for cand in candidates]
-        for model_file in self.read_preds.keys():
-            model_file_id = model_file.split('.')[-2]
-            if model_file_id not in cand_ids:
-                os.remove(model_file)
-                self.logger.info("Deleted file of non-candidate model %s", model_file)
+        for pred_file in self.read_preds.keys():
+            if pred_file not in candidates and pred_file not in self.deleted_preds:
+                # Delete prediction file
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py]: Could you please add an additional check whether there are files stored for a potential external validation and test set and delete them as well? They should be in the directories `predictions_valid` and `preditions_test`.

Comment (gui-miotto) [reply autosklearn/ensemble_builder.py]: Sure. But could you please tell me under which circumstances are those files created? Is there a test/example to which I could refer to?
When I run the test_delete_non_candidate_models, no such files are created.

Comment (mfeurer) [reply autosklearn/ensemble_builder.py]: They are created if you pass an additional `X_test` and `y_test` to the `AutoSklearnClassifier.fit()` method. You can see how they're loaded in the method `EnsembleBuilder.get_valid_test_preds()`.

Comment (gui-miotto) [reply autosklearn/ensemble_builder.py]: I tried adding `X_test` and `y_test` and the `prediction_test` folder was created, but not the `prediction_valid`.  Am I missing something?

Comment (mfeurer) [reply autosklearn/ensemble_builder.py]: >  Am I missing something?

No, actually not. Getting `prediction_valid` to be written to disk is not possible with the standard interface, so there's no need for tests for that.

**Pull request closed without merging**